### DISCUSSION TITLE
Why not use the v2 directory for version distinction?

### DISCUSSION THREAD

**Discussion by misitebao** [pelletier/go-toml#604] [Questions] [answered]
I want to know why the v2 version does not use the v2 subdirectory. Does this conform to the go module specification?

**Comment by moorereason** [comment, marked as answer]
Yes, it's a part of the Go module functionality.

**Reply by hollowaykeanho** [reply]
Supporting document: https://go.dev/blog/v2-go-modules#major-version-strategies

Quote:

```
Other strategies may keep major versions on separate branches. However, if v2+ source code is on the repository’s default branch (usually master), tools that are not version-aware — including the go command in GOPATH mode — may not distinguish between major versions.
```

**Reply by pelletier** [reply]
The use of branches for major versions is valid (see @hollowaykeanho's answer). When I started working on v2 I had to make the call of whether I wanted a `/v2` directory on `master`, or a new `v2` branch. I decided to do the latter because I found it easier to work on a separate branch: made mistakes before of opening the wrong `parser.go` file and staring at it way too long, and also generating changelogs and cutting releases is much more straightforward in the branch model. As noted in the Go blogpost, the tradeoff is that non-version aware tools would have issues with v2. That was in 2019. Given the timeline, I felt confident that all relevant tools would upgrade.
